google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/it/ITBigQueryTimeEncoderTest.java

```diff
@@ -0,0 +1,166 @@
+  public static void afterClass() {
+    if (client != null) {
+      client.close();
+    }
+    if (bigquery != null) {
+      RemoteBigQueryHelper.forceDelete(bigquery, DATASET);
+    }
+  }
+
+  @Test
+  public void TestTimeEncoding()
+      throws IOException, InterruptedException, ExecutionException,
+          Descriptors.DescriptorValidationException {
+    TableName parent = TableName.of(ServiceOptions.getDefaultProjectId(), DATASET, TABLE);
+    try (JsonStreamWriter jsonStreamWriter =
+        JsonStreamWriter.newBuilder(parent.toString(), tableInfo.getDefinition().getSchema())
+            .createDefaultStream()
+            .build()) {
+      JSONObject row = new JSONObject();
+      row.put("test_str", "Start of the day");
+      row.put(
+          "test_time_micros",
+          new JSONArray(
+              new long[] {
+                CivilTimeEncoder.encodePacked64TimeMicros(LocalTime.of(13, 14, 15, 16_000_000)),
+                CivilTimeEncoder.encodePacked64TimeMicros(LocalTime.of(23, 59, 59, 999_999_000)),
+                CivilTimeEncoder.encodePacked64TimeMicros(LocalTime.of(0, 0, 0, 0)),
+                CivilTimeEncoder.encodePacked64TimeMicros(LocalTime.of(1, 2, 3, 4_000)),
+                CivilTimeEncoder.encodePacked64TimeMicros(LocalTime.of(5, 6, 7, 8_000))
+              }));
+      row.put(
+          "test_datetime_micros",
+          new JSONArray(
+              new long[] {
+                CivilTimeEncoder.encodePacked64DatetimeMicros(
+                    LocalDateTime.of(1, 1, 1, 12, 0, 0, 0)),
+                CivilTimeEncoder.encodePacked64DatetimeMicros(
+                    LocalDateTime.of(1995, 5, 19, 10, 30, 45, 0)),
+                CivilTimeEncoder.encodePacked64DatetimeMicros(
+                    LocalDateTime.of(2000, 1, 1, 0, 0, 0, 0)),
+                CivilTimeEncoder.encodePacked64DatetimeMicros(
+                    LocalDateTime.of(2026, 3, 11, 5, 45, 12, 9_000_000)),
+                CivilTimeEncoder.encodePacked64DatetimeMicros(
+                    LocalDateTime.of(2050, 1, 2, 3, 4, 5, 6_000)),
+              }));
+      JSONArray jsonArr = new JSONArray(new JSONObject[] {row});
+      ApiFuture<AppendRowsResponse> response = jsonStreamWriter.append(jsonArr, -1);
+      Assert.assertFalse(response.get().getAppendResult().hasOffset());
+      TableResult result =
+          bigquery.listTableData(
+              tableInfo.getTableId(), BigQuery.TableDataListOption.startIndex(0L));
+      Iterator<FieldValueList> iter = result.getValues().iterator();
+      FieldValueList currentRow;
+      currentRow = iter.next();
+      assertEquals("Start of the day", currentRow.get(0).getValue());
+      assertEquals("13:14:15.016000", currentRow.get(1).getRepeatedValue().get(0).getStringValue());
+      assertEquals("23:59:59.999999", currentRow.get(1).getRepeatedValue().get(1).getStringValue());
+      assertEquals("00:00:00", currentRow.get(1).getRepeatedValue().get(2).getStringValue());
```

Comment:
Just to verify: Is the precision difference here deterministic? Or could these tests fail later due to a precision change?

Reply:
I believe it is deterministic. The time is checked that truncating to micros is equivalent to it's original time so it should return an intentional illegal argument if the precision change were to cause a problem.